### ISSUE TITLE
I want to combine road following and object detection.I want to detect a lot signs.I don’t solve it.Could anyone help me?Thank you very much.

### ISSUE THREAD

**Issue by yangrisheng** [tomMEM/Jetbot-Project#5]
I want to combine road following and object detection.I want to detect a lot signs.I don’t solve it.Could anyone help me?Thank you very much.

**Comment by tomMEM** [comment]
@yangrisheng , Hello, indeed, this is not an easy task and many different ways.
I explained under point 4 on my repository one way for the nano - it works a bit, but of course, it is not so impressive like the advertisement videos for autonomous driving.
You can find the scripts in the folder "Classification_Stop_at_RoadFollowing"; there and also under point four, I gave a few ideas on how to train for several objects (e.g.) and have a action (stop, go, turn etc.) - it is using inference/regression - it does not identify signs, persons as such - it will not create an identification box, for example.

The use of object recognition approaches (e.g. YOLO) and combination with Road Following interference model would have been better, but it seems for the nano to much - latency between object recognition and doing something is too large. But there is script in this repository (Objectfollowing and search) too.

If you like to give Classification_Stop_at_RoadFollowing  a run (can be trained for several different objects), then I still might help in case of questions. One user implemented such scripts in his repository, the link is in the folder "Classification_Stop_at_RoadFollowing".
Good luck. T

**Comment by yangrisheng** [comment]
where is this"Objectfollowing and search"?thank you for your help.Could you give me your email number,so I could contact you.Now I want to use Classification_Stop_at_RoadFollowing to finish my target.If I have some problems,I will ask u for help.thanks very much.
Good luck.

**Comment by tomMEM** [comment]
Hello @yangrisheng , the object following and free search scripts are in the primary path and explained under point two a bit: 1) live_demo-steering_tweak.ipynb and 2) TRT_object_following_tweak_object_stop.ipynb. 
They are only different regarding the environment search model that has been upgraded to TRT. 

The object recognition (e.g. water bottle, person) is based on a ssd_mobilenet_v2_coco.engine. The engine needs to be downloaded from elsewhere. The recognition is only functional with the image jetbot_JP4.3_JL1.24_PT1.30_0115.img - thus, you might need to prepare a second SD card. I do not know (Sep 2021) if somebody has converted the ssd_mobilenet to be functional with one of the last jetbot image versions.  In case of other questions (tobejpn yahoo.com). 
Best. T

**Comment by tomMEM** [comment]
Hello @yangrisheng , the object following and free search scripts are in the primary path: 1) live_demo-steering_tweak.ipynb and 2) TRT_object_following_tweak_object_stop.ipynb. They are only different regarding the environment search model that has been upgraded to TRT. 
The object recognition (e.g. water bottle, person) is based on a ssd_mobilenet_v2_coco.engine (that needs to be download from somewhere). It is only functional with the image jetbot_JP4.3_JL1.24_PT1.30_0115.img - thus, you might need to prepare a second SD card. In case of questions (tobejpn yahoo.com). 
Best. T

**Comment by tomMEM** [comment]
Still open, wrong click.

**Comment by yangrisheng** [comment]
tobejpn yahoo.com
what is this?sry, I don't understand it.

**Comment by yangrisheng** [comment]
> 你好@yangrisheng，对象跟踪和免费搜索脚本在主要路径中，并在第二点下解释一下：1）live_demo-steering_tweak.ipynb 和 2）TRT_object_following_tweak_object_stop.ipynb。
> 它们只是在升级到 TRT 的环境搜索模型上有所不同。
> 
> 对象识别（例如水瓶、人）基于 ssd_mobilenet_v2_coco.engine。引擎需要从其他地方下载。识别仅适用于图像 jetbot_JP4.3_JL1.24_PT1.30_0115.img - 因此，您可能需要准备第二张 SD 卡。我不知道（2021 年 9 月）是否有人已将 ssd_mobilenet 转换为与最新的 jetbot 映像版本之一一起使用。如果有其他问题（tobejpn yahoo.com）。
> 最好的事物。吨

Yeah.Jetpack4.5 is not support this ssd_mobilenetv2_coco.engine.And do you have used ssd_mobilenetv2_coco.engine?I intend to change my .pth model to a new mobilenetv2.engine.I using nvidia detectnet code to finish my own model.I got it.I use detectnet to gennerate a new engine.Could I use it replace the ssd_mobilenetv2_coco.engine.

**Comment by yangrisheng** [comment]
> jetbot_JP4.3_JL1.24_PT1.30_0115.img

where can I get it?jetbot_JP4.3_JL1.24_PT1.30_0115.img

**Comment by yangrisheng** [comment]
> > jetbot_JP4.3_JL1.24_PT1.30_0115.img
> 
> 哪里可以买到？jetbot_JP4.3_JL1.24_PT1.30_0115.img

https://drive.google.com/open?id=1G5nw0o3Q6E08xZM99ZfzQAe7-qAXxzHN
is this link?

**Comment by tomMEM** [comment]
Hello @yangrisheng , if you have your own model that has been created with the underlying JetPack versions then you can replace. 
It would be anyway better to have a specialized pre-trained model for your target environment.

Yes, the images are on google drive - but if you use jetbot images then they now often utilizing containers, which complicates script and installation modifications. 
The jetbot version you linked to it - I do not know, you might check "https://jetbot.org/master/software_setup/sd_card.html"
Best, T

**Comment by yangrisheng** [comment]
the JetBot containers described in this page currently target a Jetson Nano SD card image flashed with JetPack 4.4.
But I should use JetPack4.3.so the docker is not good.
And "jetbot_JP4.3_JL1.24_PT1.30_0115.img" is   "jetbot_image_v0p4p0.zip"?

**Comment by yangrisheng** [comment]
@tomMEM could you share this img?jetbot_JP4.3_JL1.24_PT1.30_0115.img

**Comment by tomMEM** [comment]
Yes, jetbot_image_v0p4p0.zip is jetbot_JP4.3_JL1.24_PT1.30_0115.img.
As a note, a manual installation on a Jetpack version you created your detectnet model might be a more save approach. 

Best. T

**Comment by yangrisheng** [comment]
@tomMEM Oh,thank you a lot.I will try it in this day.Later I will reply.

**Comment by yangrisheng** [comment]
@tomMEM Hey,I have a question.Have you retrained  ssd_mobilenetv2_coco.engine.So we can use it in own dataset.In the Nvidia tutorial,it have not this tutorial.it only has how to use this engine to detect.

**Comment by tomMEM** [comment]
@yangrisheng, For reference, ssd_mobilenetv2_coco.engines for 4.3 are linked at "https://jetbot.org/master/examples/object_following.html".

**Comment by yangrisheng** [comment]
@tomMEM I know.but do you know how to use my own dataset about object detection.

**Comment by tomMEM** [comment]
@yangrisheng, Hello, I never tried myself and can not help. Most commonly the transfer learning approach is applied - NVIDIA has quite a tutorial for that.
To start from scratch, you would need to collect and label your images with similar objects (special software can be found), and run the training approach using a deep learning layer configuration outlined in previous publications (e.g. SSD -mobile).  The nano is too small for that, it would require a workstation.

**Comment by yangrisheng** [comment]
@tomMEM Thank you.I have a own mobilenetv2 model.I will try to use it finishing my work.If it can't work,I will try to use Classification and RoadFollowing.I will need your help in that time.
Thank you very much.

**Comment by tomMEM** [comment]
@yangrisheng , looks like I kept misunderstanding the questions. But the point is, that the model will only work if similar or same dependencies, libraries (CUDA, Torch etc.) are installed on the target platform (e.g. jetbot). Since this is not the case for the old ssd_mobilenetv2_coco.engines at the newer JetPack version, it can not be applied.
Thus, to save time you have to check and test first if your model is able to run on any JetPack version after installation of all required dependencies using a minimal script.  Your workstation settings you find out with e.g. pip show or pipdeptree.
Best,T

**Comment by yangrisheng** [comment]
@tomMEM I want to combine them with my own dataset.I do not finish it.
this one is a problem.
https://forums.developer.nvidia.com/t/in-jetson-nano-using-my-own-mobilenetv2-model/188671

**Comment by tomMEM** [comment]
@yangrisheng, https://pytorch.org/docs/stable/tensors.html ?

**Comment by yangrisheng** [comment]
@tomMEM what?
